ambry-router/src/main/java/com/github/ambry/router/TtlUpdateOperation.java

```diff
@@ -355,6 +362,14 @@ private void onErrorResponse(ReplicaId replicaId, RouterException exception, boo
   private void checkAndMaybeComplete() {
     // operationCompleted is true if Blob_Authorization_Failure was received.
     if (operationTracker.isDone() || operationCompleted) {
+      if (routerConfig.routerRepairWithReplicateBLobEnabled && !operationTracker.hasSucceeded()) {
+        // the operation is failed, try to recover with ReplicateBlob
+        // if retryWithReplicateBlob returns true, it's under retry. operation is not completed yet.
+        if (retryWithReplicateBlob()) {
```

Comment:
I don't this is the best place to put this statement. The `checkAndMaybeComplete` should be restructured as such
```
if (operationTracker.isDone() || operationCompleted ) {
    if (operationTracker.hasSucceeded()) {
        operationException.set(null);
    } else if (operationTracker.getSuccessCount() > 0){
        // We have one success
        if (shouldRetryOnOneSuccess(operationException)) { // Exception that has precedency over AmbryUnavailable shouldn't be retried
            // Retry with Replicate Blob
        }
    } else if (operationTracker.maybeFailedDueToOfflineReplicas()) {
        operationException.set(new RouterException("TtlUpdateOperation failed possibly because of offline replicas", RouterErrorCode.AmbryUnavailable));
    } else if (operationTracker.hasFailedOnNotFound()) {
         operationException.set(new RouterException("TtlUpdateOperation failed because of BlobNotFound", RouterErrorCode.BlobDoesNotExist));
    }
}
```

The reason to set the order as this, is because 
1. Success already take precedence over failures
2. As long as there is one success, and the exception is not Authorization Error, or Deleted Error, or UpdateNotAllow Error, we retry
3. other errors.

Reply:
I think we have two approaches here,
1. One is the solution I'm currently using right now. Encapsulate everything into the retryWithReplicateBlob(). With that, we can easily port it to the DeleteBlob handling.

2. The other approach is as you suggested, add the code to the existing logic.
But the above "// Retry with Replicate Blob" is not a single step operation. Actually it has three stages,
a. Send ReplicateBlobRequest
b. Wait for it to be done.
c. recreate the SimpleOperationTracker and rerun the TtlUpdate.
For each sub stage, we need do different things. Also we need make sure if the TtlUpdate retry fail, don't retry again. So we need another boolean to indicate retry has been done.

if (operationTracker.isDone() || operationCompleted ) {
    if (operationTracker.hasSucceeded()) {
        operationException.set(null);
    } else if (isRunningReplicateBlob()) { // currently is executing ReplicateBlob
       // do nothing
       return;
    } else if (shouldReplicateBlob()) { // haven't do replicateBlob yet but we should call ReplicateBlob)
       // send replicate blob request
       return;
    } else if (replicateBlobCallback.isDone) {
       if (replicateBlobCallback.getException() != null) {
         // this will be the dirty part. 
         // Actually we should run the below "else if" condition "maybeFailedDueToOfflineReplicas or operationTracker.hasFailedOnNotFound" now
        // so have to move replicateBlobCallback.getException out side of this if-else-if-else chain.
        } else {
         // retry the Ttlupdate by recreating the SimpleOperationTracker etc.
          operationTracker =
              new SimpleOperationTracker(routerConfig, RouterOperation.TtlUpdateOperation, blobId.getPartition(),
                  originatingDcName, false, routerMetrics, blobId);
          ttlUpdateRequestInfos.clear();
          return true;
        }
    } else if (operationTracker.maybeFailedDueToOfflineReplicas()) {
        operationException.set(new RouterException("TtlUpdateOperation failed possibly because of offline replicas", RouterErrorCode.AmbryUnavailable));
    } else if (operationTracker.hasFailedOnNotFound()) {
         operationException.set(new RouterException("TtlUpdateOperation failed because of BlobNotFound", RouterErrorCode.BlobDoesNotExist));
    }
}

I feel embedding the retry logic into the current code logic is not clean.

Reply:
the retry logic can still be encapsulated in a single method, so we would merge `isRunningReplicateBlob` `shouldReplicaBlob` and `replicateBlobCallback.isDone`, just like the method you have now in the PR. My point is that we have to carefully structure the if statements.

Reply:
Refined.